src/spec.rs

```diff
@@ -674,25 +524,234 @@ impl Zip64CentralDirectoryEnd {
-            number_of_files,
-            central_directory_size,
-            central_directory_offset,
-        }
+
+        (
+            Zip64CDEBlock {
+                magic: Zip64CDEBlock::MAGIC,
+                record_size,
+                version_made_by,
+                version_needed_to_extract,
+                disk_number,
+                disk_with_central_directory,
+                number_of_files_on_this_disk,
+                number_of_files,
+                central_directory_size,
+                central_directory_offset,
+            },
+            extensible_data_sector,
+        )
     }
 
     pub fn write<T: Write>(self, writer: &mut T) -> ZipResult<()> {
-        self.block().write(writer)
+        let (block, comment) = self.block_and_comment();
+        block.write(writer)?;
+        writer.write_all(&comment)?;
+        Ok(())
+    }
+}
+
+pub(crate) struct CentralDirectoryEndInfo {
+    pub eocd: (Zip32CentralDirectoryEnd, u64),
+    pub eocd64: Option<(Zip64CentralDirectoryEnd, u64)>,
+
+    pub archive_offset: u64,
+}
+
+/// Finds the EOCD and possibly the EOCD64 block and determines the archive offset.
+///
+/// In the best case scenario (no prepended junk), this function will not backtrack
+/// in the reader.
+pub fn find_central_directory<R: Read + Seek>(
+    reader: &mut R,
+    archive_offset: ArchiveOffset,
+) -> ZipResult<CentralDirectoryEndInfo> {
+    const EOCD_SIG_BYTES: [u8; mem::size_of::<Magic>()] =
+        Magic::CENTRAL_DIRECTORY_END_SIGNATURE.to_le_bytes();
+
+    const EOCD64_SIG_BYTES: [u8; mem::size_of::<Magic>()] =
+        Magic::ZIP64_CENTRAL_DIRECTORY_END_SIGNATURE.to_le_bytes();
+
+    const CDFH_SIG_BYTES: [u8; mem::size_of::<Magic>()] =
+        Magic::CENTRAL_DIRECTORY_HEADER_SIGNATURE.to_le_bytes();
+
+    // Probe the file size
+    let file_length = reader.seek(io::SeekFrom::End(0))?;
+
+    // Instantiate the mandatory finder
+    let mut eocd_finder = MagicFinder::new(&EOCD_SIG_BYTES, (0, file_length));
+    let mut subfinder: Option<OptimisticMagicFinder<'static>> = None;
+
+    // Keep the last errors for cases of improper EOCD instances.
+    let mut parsing_error = None;
+
+    while let Some(eocd_offset) = eocd_finder.next_back(reader)? {
+        // Attempt to parse the EOCD block
+        let eocd = match Zip32CentralDirectoryEnd::parse(reader) {
+            Ok(eocd) => eocd,
+            Err(e) => {
+                if parsing_error.is_none() {
+                    parsing_error = Some(e);
+                }
+                continue;
+            }
+        };
+
+        // ! Relaxed (inequality) due to garbage-after-comment Python files
+        // Consistency check: the EOCD comment must terminate before the end of file
+        if eocd.zip_file_comment.len() as u64 + eocd_offset + 22 > file_length {
+            parsing_error = Some(ZipError::InvalidArchive("Invalid EOCD comment length"));
+            continue;
+        }
+
+        // Branch out for zip32
```

Comment:
Handle the case of u16::MAX files in a ZIP32 I mentioned above. This may mean changing `is_zip64()` to return a yes/no/maybe enum.

Reply:
If we support these edge cases, then the function can only return maybe/no, because it cannot verify the zip is actually zip64 locally. Changing the name of the function to `may_be_zip64` sounds like the better option.